lib/invoice_generator.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require "time"
+
+class InvoiceGenerator
+  def initialize(begin_time, end_time, save_result = false)
+    @begin_time = begin_time
+    @end_time = end_time
+    @save_result = save_result
+  end
+
+  def run
+    invoices = []
+
+    active_usage_records.group_by { |ur| ur[:project_id] }.each do |project_id, project_records|
+      project_content = {}
+      project_content[:project_id] = project_id
+      project_content[:project_name] = Project[project_id].name
+
+      project_content[:resources] = []
+      project_content[:cost] = 0
+      project_records.group_by { |pr| pr[:resource_id] }.each do |resource_id, line_items|
+        resource_content = {}
+        resource_content[:resource_id] = resource_id
+        resource_content[:resource_name] = UBID.decode(UBID.from_uuidish(resource_id)).name
+
+        resource_content[:line_items] = []
+        resource_content[:cost] = 0
+        line_items.each do |li|
+          line_item_content = {}
+          line_item_content[:location] = li[:location]
+          line_item_content[:resource_type] = li[:resource_type]
+          line_item_content[:resource_family] = li[:resource_family]
+          line_item_content[:amount] = li[:amount]
+          line_item_content[:cost] = li[:cost]
+
+          resource_content[:line_items].push(line_item_content)
+          resource_content[:cost] += line_item_content[:cost]
+        end
+
+        project_content[:resources].push(resource_content)
+        project_content[:cost] += resource_content[:cost]
+      end
+
+      if @save_result
+        Invoice.create(project_id: project_id, content: project_content)
+      else
+        invoices.push(project_content)
+      end
+    end
+
+    invoices
+  end
+
+  def active_usage_records
+    active_usage_records = UsageRecord.join(:billing_rate, id: :billing_rate_id)
+      .where { |ur| ur.initiated_at < @end_time }
```

Comment:
I think it makes sense to include the times in the comparison `<=`, what do you think?